File: Data Structures & Algorithms/count-paths/submission-5.cpp
class Solution {
    int _dfs(int r, int c, int rMax, int cMax, vector<vector<int>> &memo)
    {
        if (r > rMax || c > cMax)
            return 0;
        if(r == rMax && c == cMax) // Reached the destination
            return 1;
        if(memo[r][c] != -1)
            return memo[r][c];
        memo[r][c] = _dfs(r+1, c, rMax, cMax, memo) + _dfs(r, c+1, rMax, cMax, memo);
        return memo[r][c];
    }
public:
    int uniquePaths(int m, int n) {
        vector<vector<int>> dp(m, vector<int>(n, 0));
        dp[m-1][n-1] = 1;

        for(int r = m-1; r >= 0; r--)
            for(int c = n-1; c >= 0; c--)
            {
                if (r+1 < m)
                    dp[r][c] = dp[r+1][c];
                if (c+1 < n)
                    dp[r][c] += dp[r][c+1];
            }
        return dp[0][0];
    }
};

File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
class Solution:
    def _dfs(self, grid: List[List[str]], 
            r: int, c: int, 
            rMax: int, cMax: int, 
            visited: List[List[bool]]) -> bool:

        # check if index valid
        rInBounds = r >= 0 and r < rMax
        cInBounds = c >= 0 and c < cMax
        if not rInBounds or not cInBounds:
            return False

        # Already visited or water
        if grid[r][c] != '1' or visited[r][c] == True:
            return False

        visited[r][c] = True
        
        # explore 4 degrees of freedom
        self._dfs(grid, r-1, c, rMax, cMax, visited)
        self._dfs(grid, r+1, c, rMax, cMax, visited)
        self._dfs(grid, r, c-1, rMax, cMax, visited)
        self._dfs(grid, r, c+1, rMax, cMax, visited)

        return True

    def numIslands(self, grid: List[List[str]]) -> int:
        rMax = len(grid)
        cMax = len(grid[0])
        visited = [[False for _ in range(cMax)] for _ in range(rMax)]
        count = 0

        for r in range(0, rMax):
            for c in range(0, cMax):
                count = count + 1 if self._dfs(grid, r, c, rMax, cMax, visited) else count
        return count